File: android/src/main/java/dev/roszkowski/ScreenBrightnessMonitor.kt
package dev.roszkowski.screen_brightness_monitor

import android.content.ContentResolver
import android.content.Context
import android.database.ContentObserver
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.provider.Settings

import androidx.annotation.Keep/**
 * Monitors screen brightness and notifies registered callbacks of changes.
 *
 * @param context Android context used to access system settings.
 */
@Keep
class ScreenBrightnessMonitor(private val context: Context) {

    private val contentResolver: ContentResolver = context.contentResolver
    private val brightnessUri: Uri = Settings.System.getUriFor(Settings.System.SCREEN_BRIGHTNESS)
    private var contentObserver: ContentObserver? = null
    private var onBrightnessChanged: BrightnessCallback? = null

    /**
     * Returns the current screen brightness value (0–255).
     *
     * If the brightness cannot be read (e.g. missing permission), returns -1.
     */
    @get:Keep
    val brightness: Int
        get() = try {
            Settings.System.getInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS)
        } catch (e: Settings.SettingNotFoundException) {
            -1
        }

    /**
     * Starts observing screen brightness changes.
     *
     * The [callback] will be invoked on the main thread whenever the system
     * brightness setting changes, receiving the new brightness value (0–255).
     *
     * Only one callback can be active at a time. Calling this method again
     * replaces the previous callback (and re-registers the observer).
     */
    @Keep
    fun startObserving(callback: BrightnessCallback) {
        // Stop any existing observer first.
        stopObserving()

        onBrightnessChanged = callback

        contentObserver = object : ContentObserver(Handler(Looper.getMainLooper())) {
            override fun onChange(selfChange: Boolean) {
                super.onChange(selfChange)
                onBrightnessChanged?.onBrightnessChanged(brightness)
            }
        }

        contentResolver.registerContentObserver(brightnessUri, false, contentObserver!!)
    }

    /**
     * Stops observing screen brightness changes and removes the callback.
     */
    @Keep
    fun stopObserving() {
        contentObserver?.let {
            contentResolver.unregisterContentObserver(it)
        }
        contentObserver = null
        onBrightnessChanged = null
    }
}

File: ios/Classes/ScreenBrightnessMonitor.swift
import UIKit

/// Callback protocol for screen brightness changes.
@objc public protocol BrightnessCallback {
    /// Called when the screen brightness changes.
    ///
    /// - Parameter brightness: the new brightness value (0–255).
    @objc func onBrightnessChanged(_ brightness: Int)
}

/// Monitors screen brightness and notifies registered callbacks of changes.
@objc public class ScreenBrightnessMonitor: NSObject {
    private var observer: NSObjectProtocol?
    private var callback: BrightnessCallback?

    @objc public override init() {
        super.init()
    }

    /// Returns the current screen brightness value (0–255).
    @objc public var brightness: Int {
        return Int(UIScreen.main.brightness * 255)
    }

    /// Starts observing screen brightness changes.
    ///
    /// The callback will be invoked whenever the screen brightness changes,
    /// receiving the new brightness value (0–255).
    ///
    /// Only one callback can be active at a time. Calling this method again
    /// replaces the previous callback.
    @objc public func startObserving(callback: BrightnessCallback) {
        stopObserving()

        self.callback = callback

        observer = NotificationCenter.default.addObserver(
            forName: UIScreen.brightnessDidChangeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self = self else { return }
            self.callback?.onBrightnessChanged(self.brightness)
        }
    }

    /// Stops observing screen brightness changes and removes the callback.
    @objc public func stopObserving() {
        if let observer = observer {
            NotificationCenter.default.removeObserver(observer)
        }
        observer = nil
        callback = nil
    }

    deinit {
        stopObserving()
    }
}
